src/Mvc/Mvc.Core/src/ModelBinding/Binders/MultipartSectionModelBinder.cs

```diff
@@ -0,0 +1,108 @@
+        /// instances for reading the request body.
+        /// </param>
+        /// <param name="logger">The <see cref="ILogger"/>.</param>
+        /// <param name="options">The <see cref="MvcOptions"/>.</param>
+        public MultipartSectionModelBinder(
+            IList<IInputFormatter> formatters,
+            IHttpRequestStreamReaderFactory readerFactory,
+            ILogger<MultipartSectionModelBinder> logger,
+            MvcOptions options)
+        {
+            if (formatters == null)
+            {
+                throw new ArgumentNullException(nameof(formatters));
+            }
+
+            if (readerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(readerFactory));
+            }
+
+            _worker = new BodyModelBinderWorker(formatters, readerFactory, logger, options);
+            _logger = logger;
+        }
+
+        /// <inheritdoc />
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
+        {
+            if (bindingContext == null)
+            {
+                throw new ArgumentNullException(nameof(bindingContext));
+            }
+
+            var request = bindingContext.HttpContext.Request;
+            var requestContentType = request.ContentType;
+            if (!MediaTypeHeaderValue.TryParse(requestContentType, out var mediaType) || !mediaType.MediaType.Equals("multipart/form-data", StringComparison.OrdinalIgnoreCase))
+            {
+                // log.debug("not multipart/form-data")
+                return;
+            }
+
+            var modelBindingKey = bindingContext.BinderModelName ?? bindingContext.ModelName;
+            var form = await request.ReadFormAsync();
+            if (!form.TryGetValue(modelBindingKey, out var values) || values.Count == 0)
+            {
+                // log.debug("no value found for '{modelBindingKey}'
+                return;
+            }
+
+
+            if (!form.TryGetContentType(modelBindingKey, out var multipartContentTypeString) ||
+                multipartContentTypeString.Count == 0 ||
+                !MediaTypeHeaderValue.TryParse(multipartContentTypeString[0], out var multipartContentType))
+            {
+                // log.debug("no content-type found for '{modelBindingKey}'
+                return;
+            }
+
+            _logger?.AttemptingToBindModel(bindingContext);
+
+            var originalBody = request.Body;
+            var value = values[0];
+
+            try
+            {
+                var bytes = Encoding.UTF8.GetBytes(value);
+                request.Body = new MemoryStream(bytes);
+                request.ContentType = multipartContentType.MediaType.Value;
+
+                await _worker.ExecuteAsync(bindingContext, modelBindingKey);
+            }
+            finally
+            {
+                request.Body = originalBody;
+                request.ContentType = requestContentType;
```

Comment:
Ok, this part is wacky, you shouldn't be messing with the request like that. If you really need an HttpRequest then create a new DefaultHttpContext.

Reply:
Why change the body at all?

Reply:
> Why change the body at all?

Input formatters read the body. Unless we want to start adding special behavior in every one of those. 

> If you really need an HttpRequest then create a new DefaultHttpContext.

I don't mind doing this, but is there a reason that's preferred over changing the current request? Code hygiene?

Reply:
Assigning the body has some odd side-effects like adding a feature wrapper.

Input formatters really should be able to read an arbitrary stream.